#21/01/25
#Generates tables of the manuscript from data generated by the pipeline

########ST1 - Participants########
#Raw data only
################
########ST2 - Sampling########
#Raw data only
################
########ST3 - Genotyped SNPs########
snps <- read.csv("src/read/rawdata/101-SNP-position.csv", h = T)
snps.consensus <- read.csv("src/read/out/consensus-SNPs.csv", h = T)
snps.consensus$order <- c(1:nrow(snps.consensus))
snps <- merge(snps, snps.consensus[,c("CHROM", "POS", "order")], 
              by.x = c("Chromosome", "Position"),
              by.y = c("CHROM", "POS"), all.x = T)
colnames(snps) <- c("Locus_ID",	"Chromosome",	"Position",	
                    "Barcode order",	"Consensus barcode order")
snps <- snps[order(snps[["Barcode order"]]),]

write.csv(snps, "tables/ST3-snps.csv", quote = F, row.names = F)
rm(snps.consensus)
################
########ST4 - Barcodes########
barcodes <- read.csv("src/read/rawdata/101-genotyped-SNPS.csv", h = T)
barcodes.consensus <- read.csv("src/read/out/barcodes-consensus.csv", h = T)
uniquebarcodes <- read.csv("src/network/out/unique-barcodes.csv", h = T)[,c(1,2)]
uniquebarcodes[,2] <- "TRUE"
barcodes <- merge(barcodes, barcodes.consensus[,-which(colnames(barcodes.consensus)%in%c("Date", "Study"))], 
                  by.x = "Barcode_ID", by.y = "ID", all = T)
barcodes <- merge(barcodes, uniquebarcodes, 
                  by.x = "Barcode_ID", by.y = "ID1", all.x = T)
barcodes$score1[is.na(barcodes$score1)] <- "FALSE" 
colnames(barcodes) <- c("Barcode_ID",	"Collection type",	"Barcode", "Mixed positions",
                        "Unknown positions",	"Barcode (consensus)",	"Mixed positions (consensus)",
                        "Unknown positions (consensus)",	"Kept in figure 2")
barcodes <- barcodes[order(barcodes$Barcode_ID),]

write.csv(barcodes, "tables/ST4-barcodes.csv", quote = F, row.names = F)
rm(barcodes.consensus, uniquebarcodes)
################
########ST5 - Genomes########
#Raw data only
################
########ST6 - COI (barcodes)########
coi.barcode <- read.csv("src/coi/out/Ns-prop.csv", h = T)
coi.barcode <- coi.barcode[,c("ID", "prop.SNP")]
coi.barcode <- coi.barcode[!is.na(coi.barcode$prop.SNP),]
colnames(coi.barcode) <- c("Barcode_ID",	"Proportion of heterozygous loci")
coi.barcode <- coi.barcode[order(coi.barcode$Barcode_ID),]

write.csv(coi.barcode, "tables/ST6-coi_barcodes.csv", quote = F, row.names = F)
################
########ST7 - COI (genomes)########
coi.wgs <- read.csv("src/coi/out/Ns-prop.csv", h = T)
fws <- read.csv("src/coi/out/fws-27577-5-3750-persample.csv", h = T)
coi.wgs <- coi.wgs[,c("ID", "prop.WGS")]
coi.wgs <- coi.wgs[!is.na(coi.wgs$prop.WGS),]
coi.wgs <- merge(coi.wgs, fws, by.x = "ID", by.y = "Sample")
coi.wgs <- coi.wgs[,c("ID", "SNPs", "Fws", "prop.WGS")]

colnames(coi.wgs) <- c("Genome_ID",	"SNPs with > 5 reads over 27577 SNPs",	"Fws",	
                       "Proportion of heterozygous loci")
coi.wgs <- coi.wgs[order(coi.wgs$Genome_ID),]

write.csv(coi.wgs, "tables/ST7-coi_genomes.csv", quote = F, row.names = F)
rm(fws)
################
########ST8 - DR markers (barcodes)########
drug.phenotypes <- read.table("src/drugres/rawdata/DR-phenotypes.tsv", h = T)
drugres.barcode <- read.csv("src/drugres/out/DR_combined_WGS-barcodes_short.csv", h = T)

drug.phenotypes <- drug.phenotypes[,c("Name", "AA")]
drugres.barcode <- drugres.barcode[drugres.barcode$Type=="Barcode" & drugres.barcode$Phenotype != "Unknown",]
drugres.barcode <- merge(drugres.barcode, drug.phenotypes,
                         by = "Name")
drugres.barcode$Combinedmarker <- sapply(drugres.barcode$Combinedmarker,
                                         FUN = function(x) {
                                           paste0(unlist(strsplit(x, split = "")), collapse =",")
                                           })
drugres.barcode <- drugres.barcode[,c("Sample", "Name", "AA", "AAsen", "Combinedmarker", "Phenotype")]
colnames(drugres.barcode) <- c("Barcode_ID", "Protein",	"Position", "WT Allele",
                               "Sample Alleles",	"Genotype")
drugres.barcode <- drugres.barcode[order(drugres.barcode$Barcode_ID),]

write.table(drugres.barcode, "tables/ST8-dr_barcodes.csv", sep ="\t", quote = F, row.names = F)
rm(drug.phenotypes)
################
########ST9 - DR markers (genomes)########
drug.phenotypes <- read.table("src/drugres/rawdata/DR-phenotypes.tsv", h = T)
drugres.wgs <- read.csv("src/drugres/out/DR_combined_WGS-barcodes_short.csv", h = T)

drug.phenotypes <- drug.phenotypes[,c("Name", "AA")]
drugres.wgs <- drugres.wgs[drugres.wgs$Type=="WGS" & drugres.wgs$Phenotype != "Unknown",]
drugres.wgs <- merge(drugres.wgs, drug.phenotypes,
                         by = "Name")
drugres.wgs$Combinedmarker <- sapply(drugres.wgs$Combinedmarker,
                                         FUN = function(x) {
                                           paste0(unlist(strsplit(x, split = "")), collapse =",")
                                         })
drugres.wgs <- drugres.wgs[,c("Sample", "Name", "AA", "AAsen", "Combinedmarker", "Phenotype")]
drugres.wgs$Sample <- sub(".WGS", "", drugres.wgs$Sample)
colnames(drugres.wgs) <- c("Genome_ID", "Protein",	"Position", "WT Allele",
                               "Sample Alleles",	"Genotype")
drugres.wgs <- drugres.wgs[order(drugres.wgs$Genome_ID),]

write.table(drugres.wgs, "tables/ST9-dr_genomes.csv",  sep = "\t", quote = F, row.names = F)
rm(drug.phenotypes)
################
########ST10 - DOI########
doi <- read.csv("src/network/out/doi-metadata.csv", h = T)
doi <- doi[order(doi$Rank.in.figure.6),]
colnames(doi) <- c("Subject_ID",
                   "Estimated duration of continuous infection (days)",	
                   "Estimated start of continuous infection (YYYY-MM)",	
                   "Estimated end of continuous infection (YYYY-MM)",	"Rank in figure 6")

write.csv(doi, "tables/ST10-doi.csv", quote = F, row.names = F)
################